### PR TITLE
Adds documentation and example recommending Vec<ArrayRef> over ChunkedArray

### LINKED ISSUE #5295: Document ChunkedArray Abstractions
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The question has come up a couple of times as to why we don't have a `ChunkedArray` abstraction, we should document why we don't and what the equivalent constructions are.

**Describe the solution you'd like**

A `ChunkedArray` is really just sugar over the top of `Vec<ArrayRef>`, and is used within arrow-cpp and pyarrow for representing large in-memory datasets.

Equivalent constructions in arrow-rs would be:

* `Vec<ArrayRef>`: a fairly exact mirror to `ChunkedArray` sans some ergonomic niceties like working natively in kernels
* `impl Iterator<Item=ArrayRef>` a lazy version of a `ChunkedArray`
* `impl Stream<Item=ArrayRef>` a lazy async version of a `ChunkedArray` - this is what DataFusion uses extensively

There are also equivalent constructions using `RecordBatch` instead of `ArrayRef`.

These abstractions are strictly more flexible the pyarrow `ChunkedArray` concept, integrate better with the Rust ecosystem, and encourage users towards lazy evaluation, which has much better memory usage characteristics. 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by efredine**
# Which issue does this PR close?

Closes #5295.

# Rationale for this change
 
As described in issue.

# What changes are included in this PR?

Adds documentation and an example

# Are there any user-facing changes?

Only documentation and example.

### CODE REVIEW

File: arrow/src/lib.rs [hunk 1]
```diff
@@ -345,6 +345,67 @@
...
+//! Similar patterns can be applied at the `RecordBatch` level. For example, [DataFusion] makes
+//! extensive use of [RecordBatchStream].
+//!
+//! This approach integrates well into the Rust ecosystem, simplifies the implementation and
+//! encourages the use of performant lazy and async patterns.
+//!
+//! Aside from providing a slightly less convenient API, one other downside is the lack of support
+//! for processing compute kernels across chunked arrays. But this use case is well-supported by
+//! [DataFusion].
+//!
```
Comment (tustvold) [inline arrow/src/lib.rs:363]:
```suggestion
```

I'm not sure I agree with this statement, iterators are fairly ergonomic once you get the hang of them

File: arrow/examples/chunked_arrays.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        .unwrap(),
+    ];
+
+    // chunked_array_by_index is an array of two Vec<ArrayRef> where each Vec<ArrayRef> is a column
+    let mut chunked_array_by_index = [Vec::new(), Vec::new()];
+    for batch in &batches {
+        for (i, array) in batch.columns().iter().enumerate() {
+            chunked_array_by_index[i].push(array.clone());
+        }
+    }
```
Comment (tustvold) [inline arrow/examples/chunked_arrays.rs:57]: Collecting into separate Vec like this is unnecessary, see below

File: arrow/examples/chunked_arrays.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    // downcast and iterate over the values - column 0 is the labels and column 1 is the values
+    let labels: Vec<&str> = chunked_array_by_index[0]
+        .iter()
+        .flat_map(|x| as_string_array(x).iter())
```
Comment (tustvold) [inline arrow/examples/chunked_arrays.rs:62]: Using AsArray is typically more ergonomic

**Review by tustvold — COMMENTED** [review]
I think this example could be drastically simplified by using the iterators fully, e.g.

```
let values: Vec<f32> = batches.iter().flat_map(|x| x.column(0).as_primitive::<Float32Type>().values()).collect();

let values: Vec<&str> = batches.iter().flat_map(|x| x.column(1).as_string()).map(Option::unwrap).collect();
```

I can't see an obvious reason to collect into separate Vecs, this just adds additional overheads, and is less concise.

I think it is also worth highlighting that many use-cases won't need to cast down from the ArrayRef at all, with most kernels accept `&dyn Array`.

File: arrow/src/lib.rs [hunk 4]
```diff
@@ -345,6 +345,67 @@
 //! orchestrates the primitives exported by this crate into an embeddable query engine, with
 //! SQL and DataFrame frontends, and heavily influences this crate's roadmap.
 //!
+//! The Rust implementation does not provide the ChunkedArray abstraction implemented by the Python
```
Comment (tustvold) [inline arrow/src/lib.rs:348]: I think this documentation probably belongs at the end of these docs - https://docs.rs/arrow-array/latest/arrow_array/

Not the top-level crate documentation

**Comment by efredine** [comment]
Thanks @tustvold - I've updated the text, simplified the example as suggested and moved it to arrow_array. (I also deleted the superfluous example I had created).

File: arrow/examples/README.md [hunk 5]
```diff
@@ -21,7 +21,7 @@
 
 - [`builders.rs`](builders.rs): Using the Builder API
 - [`collect.rs`](collect.rs): Using the `FromIter` API
-- [`dynamic_types.rs`](dynamic_types.rs):
```
Comment (efredine) [inline arrow/examples/README.md:24]: Unrelated change - this example's description was missing from the README.

**Review by tustvold — APPROVED** [review]
Looks good to me, thank you

File: arrow-array/src/lib.rs [hunk 6]
```diff
@@ -161,7 +161,55 @@
...
+//! - `impl Stream<Item=ArrayRef>` a lazy async version of a `ChunkedArray`
+//!
+//! Similar patterns can be applied at the `RecordBatch` level. For example, [DataFusion] makes
+//! extensive use of [RecordBatchStream].
+//!
+//! This approach integrates well into the Rust ecosystem, simplifies the implementation and
+//! encourages the use of performant lazy and async patterns.
+//!
+//! The iterator API makes it ergonomic to work with these patterns and since most kernels accept
+//! `&dyn Array` many use-cases do not even require downcasting.
```
Comment (tustvold) [inline arrow-array/src/lib.rs:179]:
```suggestion
```
I think this is probably redundant with the section on downcasting above, but I don't fell strongly
